pkg/utils/tcp/proxy.go

```diff
@@ -0,0 +1,217 @@
+package tcp
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+type remote struct {
+	addr     string
+	mu       sync.Mutex
+	inactive bool
+}
+
+func (r *remote) isActive() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return !r.inactive
+}
+
+func (r *remote) becomeInactive() {
+	r.mu.Lock()
+	r.inactive = true
+	r.mu.Unlock()
+	log.Debug("remote become inactive", zap.String("remote", r.addr))
+}
+
+func (r *remote) dial(timeout time.Duration) error {
+	conn, err := net.DialTimeout("tcp", r.addr, timeout)
+	if err != nil {
+		return err
+	}
+	conn.Close()
+	r.mu.Lock()
+	r.inactive = false
+	r.mu.Unlock()
+	return nil
+}
+
+type Proxy struct {
```

Comment:
actually this is a forwarder

Reply:
Proxy is ok I think 😁

Reply:
Then maybe you can remove the "forwarder", since it was created exactly for this purpose (but without any implementation XD)